plugin/iplookup/cache.go

```diff
@@ -0,0 +1,119 @@
+package iplookup
+
+import (
+	"net/http"
+	"time"
+)
+
+// Server the cache
+func (ipl *IPLookup) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	lookup := r.URL.String()
+	if len(lookup) > 0 && lookup[0] == '/' {
+		lookup = lookup[1:]
+	}
+
+	ipl.Lock()
+	entry, _ := ipl.cacheLookup[lookup]
+	ipl.Unlock()
+
+	if entry == nil {
+		http.NotFound(w, r)
+		return
+	}
+
+	// Write the entry
+	w.Write([]byte(entry.value))
+}
+
+func (ipl *IPLookup) addCache(name, value string) {
+
+	ipl.Lock()
+
+	// See if we know about this entry
+	entry, found := ipl.cacheLookup[name]
+	if found {
+
+		// Update the expires time
+		entry.expires = time.Now().Add(ipl.maxDuration)
+
+		// We only need to modify if we're not the head already
+		if entry.next != nil {
+
+			// We are the tail, we know we're not the head
+			if entry.prev == nil {
+				ipl.tail = entry.next
+				entry.next.prev = nil
+			} else {
+				// We know we're not the tail and not the head
+				entry.prev.next = entry.next
+				entry.next.prev = entry.prev
+			}
+
+			ipl.head.next = entry
+			entry.prev = ipl.head
+			entry.next = nil
+			ipl.head = entry
+
+		}
+
+	} else {
+
+		// Doesn't exist, add it to the cache
+		entry = &cacheEntry{
+			name:    name,
+			value:   value,
+			expires: time.Now().Add(ipl.maxDuration),
+
+			next: nil,
+			prev: ipl.head,
+		}
+
+		if ipl.head != nil {
+			ipl.head.next = entry
+		} else {
+			ipl.tail = entry // Head is nil, so is tail, set it
+		}
+		ipl.head = entry
+		ipl.cacheLookup[name] = entry
+
+		ipl.size++
+
+	}
+
+	ipl.Unlock()
+
+}
+
+func (ipl *IPLookup) cleanCache() {
+
+	ipl.Lock()
+
+	now := time.Now()
+
+	for ipl.tail != nil {
```

Comment:
Why not:

```
for ipl.tail != nil && (ipl.maxDuration > 0 && now.After(ipl.tail.expires)) || (ipl.maxEntries > 0 && ipl.size > ipl.maxEntries) {
// Update the tail next.prev
  if ipl.tail.next != nil {
    ipl.tail.next.prev = nil
  }

  // Delete from the lookup
  delete(ipl.cacheLookup, ipl.tail.name)

  // Advance the tail pointer
  ipl.tail = ipl.tail.next

  ipl.size--
}
```

Then there's no need for the break and you save an indent level.

Reply:
even better to keep `now` local:
```
for now := time.Now(); ipl.tail != nil && (ipl.maxDuration > 0 && now.After(ipl.tail.expires)) || (ipl.maxEntries > 0 && ipl.size > ipl.maxEntries); {
// Update the tail next.prev
  if ipl.tail.next != nil {
    ipl.tail.next.prev = nil
  }

  // Delete from the lookup
  delete(ipl.cacheLookup, ipl.tail.name)

  // Advance the tail pointer
  ipl.tail = ipl.tail.next

  ipl.size--
}
```

Reply:
It's completely for readability. I (personally) find it much easier to read and figure out what's going on in the function. What you suggested would keep everything in the for loop statement true, but it becomes quite long and it actually needs another set of parenthesis to be evaluated correctly. I would argue that it would compile down to pretty much same code as well.